components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipPresenceAgentListener.java

```diff
@@ -179,6 +187,64 @@ public void processSubscribe(RequestEvent requestEvent,
         }
     }
 
+    public void processMessage(RequestEvent requestEvent, ServerTransaction serverTransactionId) {
+        Request request = requestEvent.getRequest();
+        try
+        {
+            LOG.debug("SipPresenceAgentListener: Received an MESSAGE request");
+            LOG.debug("SipPresenceAgentListener retrieved MESSAGE: {}", requestEvent.getRequest());
+
+            Response response = null;
+            MessageFactory messageFactory = sipPresenceAgent.getConfiguration().getMessageFactory();
+            boolean contentTypeSupported = false;
+
+            SipProvider sipProvider = (SipProvider) requestEvent.getSource();
+            String contentType = request.getHeader("Content-Type").toString();
+
+            if(contentType == null)
+            {
+                response = messageFactory.createResponse(400, request); // bad request
+                sipProvider.sendResponse(response);
+                return;
+            }
+            else if(contentType.contains("text/plain"))
+            {
+                contentTypeSupported = true;
+                messageQueue.add((String) request.getContent());
+            }
+
+            //send OK or unsupported based the retrieved content type
+            if(!contentTypeSupported)
+            {
+                response = messageFactory.createResponse(415, request); // unsupported media type
+            }
+            else
+            {
+                response = messageFactory.createResponse(200, request); // OK
+            }
+            sipProvider.sendResponse(response);
+        }
+        catch( Exception e)
+        {
+            LOG.debug("Exception thrown during processing Message request in SipPresenceAgentListener");
+        }
+    }
+
+    public boolean hasRetrievedMessage()
+    {
+        return !messageQueue.isEmpty();
+    }
+
+    public String getRetrievedMessage() throws NoSuchElementException
```

Comment:
@nikvaessen instead of just returning the message body as a string, could you create an Exchange object and populate the "in" message body and various headers?